### PR TITLE
Add inject_points for projecting external points onto the network

### PR DESCRIPTION
**Pull request by jg-codes**
## Summary
This PR adds `inject_points`, a focused helper for projecting external Point features onto the nearest network LineString and splitting at the projected location.

It follows the scoped discussion in #304: this PR includes only `inject_points`, its public export, and the API reference entry. The separate node-attribute attachment helper is intentionally left out for a future discussion.

## What Changed
- Add `neatnet.inject_points`
- Add tests for on-line and near-line projection, `snap_radius` filtering, CRS reprojection, empty inputs, and multiple points on one edge
- Add `inject_points` to the API reference

## Validation
- `git diff --check upstream/main...HEAD` passes
- Local focused pytest execution was blocked by missing `libpysal` in this local environment; GitHub Actions should run the project environment

AI assistance disclosure: Prepared with AI assistance; final scope, structure, and review decisions were made by JG.

### CODE REVIEW

File: neatnet/nodes.py [hunk 1]
```diff
@@ -343,6 +347,78 @@ def induce_nodes(streets: gpd.GeoDataFrame, *, eps: float = 1e-4) -> gpd.GeoData
...
+
+    See Also
+    --------
+    induce_nodes : symmetric case where the new node comes from a LineString endpoint.
+    split : underlying split-at-points primitive.
+    """
+    if streets.crs is None or points.crs is None:
+        raise ValueError("Both streets and points must have a CRS set.")
+    if points.crs != streets.crs:
+        points = points.to_crs(streets.crs)
```
Comment (martinfleis) [inline neatnet/nodes.py]: I would do this here as we do in `neatify` - check and if it does not match raise. Let users fix it, rather than implicitly re-project.

File: neatnet/nodes.py [hunk 2]
```diff
@@ -343,6 +347,78 @@ def induce_nodes(streets: gpd.GeoDataFrame, *, eps: float = 1e-4) -> gpd.GeoData
...
+
+    line_geoms = streets.geometry.values
+    point_geoms = points.geometry.values
+    if len(point_geoms) == 0 or len(line_geoms) == 0:
+        return streets.copy()
+
+    # Nearest line per input point (sindex-accelerated)
+    input_idx, line_idx = streets.sindex.nearest(point_geoms, return_all=False)
+    distances = shapely.distance(line_geoms[line_idx], point_geoms[input_idx])
```
Comment (martinfleis) [inline neatnet/nodes.py]: `sindex.nearest` can directly return distances - see [docs](https://geopandas.org/en/stable/docs/reference/api/geopandas.sindex.SpatialIndex.nearest.html)

**Review by martinfleis — COMMENTED** [review]
A first pass without checking the tests yet.

Out of curiosity, what happens on your real use case when you use directly `split` with `eps` equal to what you use as `snap_radius` here? I have a feeling that you might actually get the same outcome as with `inject_points`.

**Comment by jg-codes** [comment]
> A first pass without checking the tests yet.
> 
> Out of curiosity, what happens on your real use case when you use directly `split` with `eps` equal to what you use as `snap_radius` here? I have a feeling that you might actually get the same outcome as with `inject_points`.

Thanks. Both changes are pushed.

On `split(eps=snap_radius)`: I ran it on my actual data, the full German federal waterway network (VerkNet, ~3,000 segments) with RIS locks, ports and bridges as the points. The first figure is the offset of each feature from the nearest fairway centreline.

germany_poi_offset_ecdf

For bridges (median 2 m) and locks (median 1 m) you're right: at any sensible eps, split and inject_points give the same node. The difference is ports. They sit in harbour basins, median 120 m off the fairway, with almost none within 5 m. There split(eps) drags the fairway out to the port and leaves a large kink, while inject_points projects to the foot and keeps the line straight.

germany_split_vs_inject_geo

A perpendicular foot may not be quite right for a port either, since a port connects through its harbour entrance and logistics activities can be on the piers and elsewhere, not a straight drop onto the main line. For those you may need a connector edge, which I kept out of scope. 

Does that scoping match your expectation?

**Comment by martinfleis** [comment]
Ah, great! Thanks a lot. We should make sure this comes clear through the docs, both docstrings of both functions and in the user guide notebook. The fact that `split` forces line to go through the point while `inject_points` keeps the original linework, just subdivided.

**Comment by jg-codes** [comment]
> Ah, great! Thanks a lot. We should make sure this comes clear through the docs, both docstrings of both functions and in the user guide notebook. The fact that `split` forces line to go through the point while `inject_points` keeps the original linework, just subdivided.

I've added the distinction to the docs. Let me know if anything else comes up your mind.

File: neatnet/nodes.py [hunk 3]
```diff
@@ -53,7 +53,15 @@ def split(
     *,
     eps: float = 1e-4,
 ) -> gpd.GeoSeries | gpd.GeoDataFrame:
-    """Split lines on new nodes.
+    """Split lines at the given points, snapping each line through the point.
+
+    Each point is snapped onto the nearest line within ``eps`` (via
```
Comment (martinfleis) [inline neatnet/nodes.py]: Line is snapped onto point, not point onto line, no?

File: neatnet/nodes.py [hunk 4]
```diff
@@ -343,6 +360,104 @@ def induce_nodes(streets: gpd.GeoDataFrame, *, eps: float = 1e-4) -> gpd.GeoData
...
+    *,
+    snap_radius: float | None = None,
+    eps: float = 1e-4,
+) -> gpd.GeoDataFrame:
+    """Project external points onto the nearest line and split it at the foot.
+
+    Each input point within ``snap_radius`` of the network is projected onto
+    its *nearest* LineString in ``streets`` -- at the foot of the
+    perpendicular, via linear referencing -- and that line is split at the
+    projected location. The new node therefore lands exactly *on* the original
```
Comment (martinfleis) [inline neatnet/nodes.py]: Not exactly as we're snapping coordinates into a floating point precision grid but almost exactly. I would avoid using the word `exactly` here as the original and new lines will not overlap 100%.

File: neatnet/nodes.py [hunk 5]
```diff
@@ -343,6 +360,104 @@ def induce_nodes(streets: gpd.GeoDataFrame, *, eps: float = 1e-4) -> gpd.GeoData
...
+        )
+
+    line_geoms = streets.geometry.values
+    point_geoms = points.geometry.values
+    if len(point_geoms) == 0 or len(line_geoms) == 0:
+        return streets.copy()
+
+    # Nearest line per input point, with distance (sindex-accelerated)
+    (input_idx, line_idx), distances = streets.sindex.nearest(
+        point_geoms, return_all=False, return_distance=True
```
Comment (martinfleis) [inline neatnet/nodes.py]: If the `snap_radius` is set, we could pass it here as `max_distance` to speed this up.

File: neatnet/nodes.py [hunk 6]
```diff
@@ -343,6 +360,104 @@ def induce_nodes(streets: gpd.GeoDataFrame, *, eps: float = 1e-4) -> gpd.GeoData
...
+
+    Parameters
+    ----------
+    streets : geopandas.GeoDataFrame
+        LineString network. CRS must be set.
+    points : geopandas.GeoDataFrame
+        Point features to project onto ``streets``. Must share the same CRS as
+        ``streets`` (a mismatch raises rather than reprojecting implicitly).
+        Only the geometries are used; point attributes are not carried over.
+    snap_radius : float | None = None
```
Comment (martinfleis) [inline neatnet/nodes.py]: Not sure if `snap_radius` is the right name as we're technically not snapping. Maybe keep just `radius`?

File: neatnet/nodes.py (hunk 3 above)
Comment (jg-codes) [reply neatnet/nodes.py]: Right; that sentence had it backwards. The same inversion sat in the eps parameter description and in the pre-existing _snap_n_split helper docstring ("Snap point to edge"), so fixed those in the same pass.

File: neatnet/nodes.py (hunk 4 above)
Comment (jg-codes) [reply neatnet/nodes.py]: reworded to "up to floating-point precision",

File: neatnet/nodes.py (hunk 5 above)
Comment (jg-codes) [reply neatnet/nodes.py]: Done; radius now goes straight into sindex.nearest as max_distance (distances <= radius)

File: neatnet/nodes.py (hunk 6 above)
Comment (jg-codes) [reply neatnet/nodes.py]: renamed to `radius` throughout (signature, tests, docs).

File: neatnet/nodes.py (hunk 5 above)
Comment (jg-codes) [reply neatnet/nodes.py]: However, two details I checked while swapping:

-  the bound is inclusive at the boundary, same as the old distances <= radius, and
-  `radius=0` (meaning: inject only points already on a line) can't be expressed as max_distance, which geopandas requires to be positive. Hence, this case keeps a small distance-filter fallback. 

Both behaviors are pinned by new test cases: `test_inject_points_projects_to_target_line` case with a point at distance exactly `radius` (the bound stays inclusive, matching the previous `distances <= radius`):

https://github.com/uscuni/neatnet/blob/5c4f385/neatnet/tests/test_nodes.py#L1341-L1345

and `test_inject_points_zero_radius_keeps_only_on_line_points` — `radius=0` still means "on-line points only" via a small distance-filter fallback, since `max_distance` must be positive:

https://github.com/uscuni/neatnet/blob/5c4f385/neatnet/tests/test_nodes.py#L1372-L1381

File: neatnet/nodes.py [hunk 7]
```diff
@@ -435,15 +437,20 @@ def inject_points(
...
-
-    if snap_radius is not None:
-        keep = distances <= snap_radius
-        input_idx = input_idx[keep]
-        line_idx = line_idx[keep]
+    if radius == 0:
+        # geopandas rejects ``max_distance=0``: filter to on-line points
+        (input_idx, line_idx), distances = streets.sindex.nearest(
+            point_geoms, return_all=False, return_distance=True
+        )
```
Comment (martinfleis) [inline neatnet/nodes.py:445]: You can still include some small max_distance here to speed up the search.

Comment (jg-codes) [reply neatnet/nodes.py:445]: Good catch; that branch was still unbounded. Fixed in c91bae9.

As `max_distance=0` is rejected by geopandas, but a *positive* bound is fine and
doesn't change the result here: I pass `max_distance=1e-9` and keep the
`distances == 0` filter. Every on-line point has distance exactly `0 <= 1e-9`,
so it's always returned; off-line points are now pruned during the tree
traversal instead of being fully resolved and then discarded.

```python
(input_idx, line_idx), distances = streets.sindex.nearest(
    point_geoms, return_all=False, return_distance=True, max_distance=1e-9
)
on_line = distances == 0
```

(
`radius=None` stays unbounded as documented as injecting
every point regardless of distance from the network.
I kept `nearest` + `== 0` rather than switching the zero case to
`sindex.query(..., predicate="intersects")`, since `intersects` would inject
into *every* coincident line rather than just the nearest one and would lean on
exact floating-point coincidence. Happy to switch to the query form if you'd
prefer it.
)

**Review by martinfleis — APPROVED** [review]

**Review by jGaboardi — APPROVED** [review]
While I am not thrilled about the use of Claude (or any other LLM) in actual commits, this is indeed a nice and useful contribution. Thank you very much for it @jg-codes.

**Comment by jg-codes** [comment]
> While I am not thrilled about the use of Claude (or any other LLM) in actual commits, this is indeed a nice and useful contribution. Thank you very much for it @jg-codes.

@martinfleis  @jGaboardi  thank you for your sharp review, your discussion points, and support. It helped me in growing my understanding.
